File: cpp/tests/spsc_ring_buffer_test.cpp
#include "test_harness.hpp"
#include "spsc_ring_buffer.hpp"
#include <thread>
#include <cstdint>

int main() {
    TestHarness t;
    std::cout << "SpscRingBufferTest\n";

    // Functional: fills exactly to capacity, then rejects; FIFO order preserved.
    {
        SpscRingBuffer<int, 4> rb;
        t.check("push #1 succeeds", rb.push(10));
        t.check("push #2 succeeds", rb.push(20));
        t.check("push #3 succeeds", rb.push(30));
        t.check("push #4 succeeds (buffer now full at capacity)", rb.push(40));
        t.check("push #5 fails while full", !rb.push(50));

        int out = -1;
        t.check("pop #1 succeeds", rb.pop(out));
        t.checkEquals("pop #1 returns items in FIFO order", 10, out);
        t.check("pop #2 succeeds", rb.pop(out));
        t.checkEquals("pop #2 returns items in FIFO order", 20, out);

        // Freed a slot: push should succeed again, demonstrating wraparound.
        t.check("push after pop succeeds (wraparound)", rb.push(50));

        t.check("pop #3 succeeds", rb.pop(out));
        t.checkEquals("pop #3 returns items in FIFO order", 30, out);
        t.check("pop #4 succeeds", rb.pop(out));
        t.checkEquals("pop #4 returns items in FIFO order", 40, out);
        t.check("pop #5 succeeds", rb.pop(out));
        t.checkEquals("pop #5 returns the wrapped-around item", 50, out);
        t.check("pop fails once empty", !rb.pop(out));
    }

    // Concurrent: one producer, one consumer, verify no lost/duplicated
    // items via a checksum -- this is the property the release/acquire
    // pairing on head/tail is actually there to guarantee.
    {
        constexpr int total = 200000;
        static SpscRingBuffer<int, 1024> rb;

        std::thread producer([&] {
            for (int i = 0; i < total; ++i) {
                while (!rb.push(i)) std::this_thread::yield();
            }
        });

        long long sum = 0;
        std::thread consumer([&] {
            int item;
            for (int i = 0; i < total; ++i) {
                while (!rb.pop(item)) std::this_thread::yield();
                sum += item;
            }
        });

        producer.join();
        consumer.join();

        long long expected = static_cast<long long>(total) * (total - 1) / 2;
        t.checkEquals("concurrent producer/consumer: checksum matches (no lost/dup items)",
                      expected, sum);
    }

    return t.summary("SpscRingBufferTest");
}

File: java/tests/SpscRingBufferTest.java
/** Same suite as cpp/tests/spsc_ring_buffer_test.cpp -- see that file for
 *  the C++ side of this exact comparison. */
public class SpscRingBufferTest {

    static int run() throws InterruptedException {
        TestHarness t = new TestHarness();
        System.out.println("SpscRingBufferTest");

        // Functional: fills exactly to capacity, then rejects; FIFO order preserved.
        SpscRingBuffer<Integer> rb = new SpscRingBuffer<>(4);
        t.check("offer #1 succeeds", rb.offer(10));
        t.check("offer #2 succeeds", rb.offer(20));
        t.check("offer #3 succeeds", rb.offer(30));
        t.check("offer #4 succeeds (buffer now full at capacity)", rb.offer(40));
        t.check("offer #5 fails while full", !rb.offer(50));

        t.checkEquals("poll #1 returns items in FIFO order", 10, rb.poll());
        t.checkEquals("poll #2 returns items in FIFO order", 20, rb.poll());

        // Freed two slots: offers should succeed again (wraparound).
        t.check("offer after poll succeeds (wraparound)", rb.offer(50));
        t.check("second offer after poll succeeds (wraparound)", rb.offer(60));

        t.checkEquals("poll #3 returns items in FIFO order", 30, rb.poll());
        t.checkEquals("poll #4 returns items in FIFO order", 40, rb.poll());
        t.checkEquals("poll #5 returns the first wrapped-around item", 50, rb.poll());
        t.checkEquals("poll #6 returns the second wrapped-around item", 60, rb.poll());
        t.checkEquals("poll returns null once empty", null, rb.poll());

        // Concurrent: one producer, one consumer, verify no lost/duplicated
        // items via a checksum -- the property the volatile-backed
        // head/tail publish/subscribe pattern is there to guarantee.
        final int total = 200_000;
        SpscRingBuffer<Integer> concurrentRb = new SpscRingBuffer<>(1024);
        long[] sum = {0};

        Thread producer = new Thread(() -> {
            for (int i = 0; i < total; i++) {
                while (!concurrentRb.offer(i)) Thread.onSpinWait();
            }
        });
        Thread consumer = new Thread(() -> {
            for (int i = 0; i < total; i++) {
                Integer item;
                while ((item = concurrentRb.poll()) == null) Thread.onSpinWait();
                sum[0] += item;
            }
        });
        producer.start();
        consumer.start();
        producer.join();
        consumer.join();

        long expected = (long) total * (total - 1) / 2;
        t.checkEquals("concurrent producer/consumer: checksum matches (no lost/dup items)",
                expected, sum[0]);

        return t.summary("SpscRingBufferTest");
    }

    public static void main(String[] args) throws InterruptedException {
        System.exit(run());
    }
}
